Added agent option to http transport

RAWeber opened the pull request:
Provide option to set a custom agent to use with node-fetch.

https://www.npmjs.com/package/node-fetch#custom-agent

This will allow users to specify any networking related options. For example, users can provide client certificates to hit mTLS endpoints, or CA certificates to use self-signed certificates.

Will use default agent if no agent option passed.

adriancole commented:
let's default this to the current version of the transport. ex

`zipkin-transport-http/0.19.0`

While not here, we've had quite a discussion elsewhere on it. https://github.com/openzipkin/zipkin-reporter-java/issues/142

RAWeber commented:
Based on node-fetches [docs](https://nodejs.org/api/http.html#http_class_http_agent) and the nodejs http.Agent [docs](https://nodejs.org/api/http.html#http_class_http_agent) which it refers, I believe the user-agent referred to in: openzipkin/zipkin-reporter-java#142 would simply be a matter of updating the header[user-agent] property, while the Agent option I've added would refer to an http.Agent.

The http.agent would be responsible for managing connection persistence and reuse for HTTP clients, such as sockets, certs, etc.

I believe `zipkin-transport-http/0.19.0` could just be added to the default headers. I can add that to this commit if you would like, but it is not necessarily related to the agent parameter.

The current default uses the http.globalAgent, which has the base [defaults](https://nodejs.org/docs/latest-v7.x/api/http.html#http_new_agent_options) (this would reflect current behavior where no http.agent is being used)

I've also updated the `index.d.ts` file to reflect the proper type of the agent option.

adriancole commented:
sorry wrong agent I guess!

adriancole commented on packages/zipkin-transport-http/index.d.ts:
```diff
@@ -1,12 +1,17 @@
     jsonEncoder?: JsonEncoder,
     httpTimeout?: number,
     headers?: { [name: string]: any },
+    agent?: Agent | (() => Agent),
```
is `(() => Agent)` the same effect as null?

Let's add typescript and javascript tests for this parameter to make sure someone later doesn't accidentally break what you need.

RAWeber replied:
`(() => Agent)` is the equivalent of a function which returns an Agent

This follows node-fetch which will accept an HTTP(S) Agent or a function which returns a HTTP(S) Agent

RAWeber commented on package.json:
```diff
@@ -14,7 +14,7 @@
     "lint": "npm-run-all --parallel lint:*",
     "test:es": "npm run lerna-test",
-    "test:ts": "tsr packages/zipkin/test-types/*.test.ts --noAnnotate --libDeclarations && mocha --opts test/mocha-types.opts",
+    "test:ts": "tsr packages/*/test-types/*.test.ts --noAnnotate --libDeclarations && mocha --opts test/mocha-types.opts",
```
Updated this to process `test-type/*.test.ts` tests in all packages. This reflects the second mocha command  which   uses `test/mocha-types.opts` and attempts to run test-type tests in all packages.

RAWeber commented:
Build failed on the Kafka Transport. The only change which could affect this (at least based on my current understanding of the project) is the change to the `test:ts` however, it failed on the `test:es` script, so I do not believe this was the cause.

Is there possibly another issue going on? Or an intermittent issue unrelated to these changes?

adriancole commented on packages/zipkin-transport-http/test-types/HttpLogger.test.ts:
```diff
@@ -0,0 +1,37 @@
+  });
+
+  it('should accept Http(s) Agent or function which returns Agent', () => {
+    const agents = [new HttpAgent(), new HttpsAgent(), () => new HttpAgent(), () => new HttpsAgent()];
```
so this test makes sure it doesn't crash in the constructor. am I understanding that right?

I wonder if it would crash later. It might be confusing to have a test without assertions I mean.. wdyt?

RAWeber replied:
Ya it's testing the constructor. I've added an expect statement to confirm `this.agent` is properly being set and defaulting to `null` if it is not provided.

Beyond that, it is only being forwarded to the fetch call. Ideally I could confirm that fetch is being called with the agent, however due to the `node-fetch` implementation and `sinon` restrictions, it doesn't seem feasible to perform such a test without some hacky solutions.

adriancole commented:
kicked build in case it is a flake

jcchavezs commented:
Sorry I am too late for this discussion.

Out of curiosity: is this agent is also the same agent supported in axios:
https://github.com/axios/axios/blob/master/README.md#request-config

In addition, I feel like we haven't added any sort of documentation about
this new feature. Could we add at least in tsdoc about this new parameter?

In any case great work @RAWeber!


man. 19. aug. 2019, 14:30 skrev Adrian Cole <notifications@github.com>:

> Merged #439 <https://github.com/openzipkin/zipkin-js/pull/439> into
> master.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/openzipkin/zipkin-js/pull/439?email_source=notifications&email_token=AAXOYAXTNBN2J56ITMIQ56TQFKG5NA5CNFSM4ILFMEV2YY3PNVWWK3TUL52HS4DFWZEXG43VMVCXMZLOORHG65DJMZUWGYLUNFXW5KTDN5WW2ZLOORPWSZGOTD2AT3Q#event-2566130158>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAXOYATHN5JGF7RK5GNIUZLQFKG5NANCNFSM4ILFMEVQ>
> .
>

RAWeber commented:
@jcchavezs looks to be the same agent supported by Axios, as both just accept a Nodejs http(s) Agent.

And I will look into adding some documentation this week!
